### PR TITLE
remove `KEEP_IN_LOCAL_TCX` flag

### LINKED ISSUE #70285: Replace calls to has_local_value with needs_infer.
The only difference between the two is that "fresh" variables (a predecessor of the "canonical query" system) are considered `needs_infer`, despite not really taking part in inference.

I think that even without changing/replacing the "freshening" system at all, we can make "fresh" variables not `needs_infer`, removing the need for `TypeFlags::KEEP_IN_LOCAL_TCX` and `has_local_value` (which only existed for the `'gcx` vs `'tcx` split anyway).

If we want to be more explicit, we could rename `needs_infer` to `is_local_to_inference_context` or something similar, but either way I think it makes more sense than "local value".

cc @nikomatsakis @matthewjasper @Zoxc 

This issue has been assigned to @lcnr via [this comment](https://github.com/rust-lang/rust/issues/70285#issuecomment-606851898).

### PR DESCRIPTION
**Pull request by lcnr**
closes #70285

I did not rename `needs_infer` here as this complex enough as is.
Will probably open a followup for that.

r? @eddyb

### CODE REVIEW

File: src/librustc_infer/infer/canonical/canonicalizer.rs [hunk 1]
```diff
@@ -524,7 +524,7 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
         // Once we have canonicalized `out_value`, it should not
         // contain anything that ties it to this inference context
         // anymore, so it should live in the global arena.
-        debug_assert!(!out_value.has_type_flags(TypeFlags::KEEP_IN_LOCAL_TCX));
+        debug_assert!(!out_value.has_type_flags(TypeFlags::NEEDS_INFER));
```
Comment (eddyb) [inline src/librustc_infer/infer/canonical/canonicalizer.rs:527]: `.has_type_flags(TypeFlags::NEEDS_INFER)` -> `.needs_infer()`.

File: src/librustc_typeck/check/writeback.rs [hunk 2]
```diff
@@ -579,7 +579,7 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         T: TypeFoldable<'tcx>,
     {
         let x = x.fold_with(&mut Resolver::new(self.fcx, span, self.body));
-        if cfg!(debug_assertions) && x.has_local_value() {
+        if cfg!(debug_assertions) && x.needs_infer() {
             span_bug!(span.to_span(self.fcx.tcx), "writeback: `{:?}` is a local value", x);
```
Comment (eddyb) [inline src/librustc_typeck/check/writeback.rs:583]: `is a local value` -> `has inference variables`

File: src/librustc_typeck/check/writeback.rs [hunk 3]
```diff
@@ -399,7 +399,7 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         assert_eq!(fcx_tables.hir_owner, self.tables.hir_owner);
 
         for (&def_id, c_sig) in fcx_tables.user_provided_sigs.iter() {
-            if cfg!(debug_assertions) && c_sig.has_local_value() {
+            if cfg!(debug_assertions) && c_sig.needs_infer() {
                 span_bug!(
                     self.fcx.tcx.hir().span_if_local(def_id).unwrap(),
                     "writeback: `{:?}` is a local value",
```
Comment (eddyb) [inline src/librustc_typeck/check/writeback.rs:405]: Same here (check the entire file, or maybe even the entire compiler).

Comment (lcnr) [reply src/librustc_typeck/check/writeback.rs:405]: jup

File: src/librustc_trait_selection/traits/select.rs [hunk 4]
```diff
@@ -1181,7 +1181,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         // If there are any e.g. inference variables in the `ParamEnv`, then we
```
Comment (eddyb) [inline src/librustc_trait_selection/traits/select.rs:1181]: You can remove "e.g." now.

File: src/librustc_trait_selection/traits/mod.rs [hunk 5]
```diff
@@ -259,7 +259,7 @@ fn do_normalize_predicates<'tcx>(
                 return Err(ErrorReported);
             }
         };
-        if predicates.has_local_value() {
+        if predicates.needs_infer() {
             // FIXME: shouldn't we, you know, actually report an error here? or an ICE?
             Err(ErrorReported)
```
Comment (eddyb) [inline src/librustc_trait_selection/traits/mod.rs:264]: cc @nikomatsakis This is a scary FIXME comments :open_mouth:. Should we add a `delay_span_bug`?

Comment (lcnr) [reply src/librustc_trait_selection/traits/mod.rs]: added a `delay_span_bug`. Should be a relevant error msg afaik
> encountered inference variables after `fully_resolve`

File: src/librustc_middle/ty/relate.rs [hunk 6]
```diff
@@ -510,7 +510,7 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
     let tcx = relation.tcx();
 
     let eagerly_eval = |x: &'tcx ty::Const<'tcx>| {
-        if !x.val.has_local_value() {
+        if !x.val.needs_infer() {
             return x.eval(tcx, relation.param_env()).val;
```
Comment (eddyb) [inline src/librustc_middle/ty/relate.rs:514]: cc @oli-obk @varkor This seems unnecessary, `ty::Const::eval` handles `needs_infer` itself.

File: src/librustc_middle/ty/sty.rs [hunk 7]
```diff
@@ -2404,7 +2403,7 @@ impl<'tcx> Const<'tcx> {
             let param_env_and_substs = param_env.with_reveal_all().and(substs);
 
             // Avoid querying `tcx.const_eval(...)` with any e.g. inference vars.
```
Comment (eddyb) [inline src/librustc_middle/ty/sty.rs:2405]: Same "e.g." here that can be removed.

File: src/librustc_middle/ty/sty.rs [hunk 8]
```diff
@@ -2424,7 +2423,7 @@ impl<'tcx> Const<'tcx> {
                 // when the expression doesn't depend on any parameters.
                 // FIXME(eddyb, skinny121) pass `InferCtxt` into here when it's available, so that
                 // we can call `infcx.const_eval_resolve` which handles inference variables.
-                if substs.has_local_value() {
+                if substs.needs_infer() {
```
Comment (eddyb) [inline src/librustc_middle/ty/sty.rs:2426]: That's an "e.g." here too. Maybe search for "e.g. inference variables" across the entire codebase?

**Review by eddyb — APPROVED** [review]
LGTM

**Review by nikomatsakis — APPROVED** [review]

**Comment by lcnr** [comment]
should once again be ready

**Pull request merged**